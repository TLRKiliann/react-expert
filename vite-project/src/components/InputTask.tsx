type PropsValue = {
	valueToRetrieve: {
		firstName: string,
		age: number,
		admin: boolean
	};
}

const InputTask = (props: PropsValue): JSX.Element => {
	return (
		<>
			<h1>Hello InputTask</h1>
			<h1 style={{color: "deepskyblue"}}>{props.valueToRetrieve.firstName}</h1>
			<h1 style={{color: "deepskyblue"}}>{props.valueToRetrieve.age}</h1>
			<h1 style={{color: "deepskyblue"}}>{props.valueToRetrieve.admin ? `Yes admin` : `not admin`}</h1>
		</>
	)
}

export default InputTask;