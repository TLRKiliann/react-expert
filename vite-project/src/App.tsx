import React, { useState } from 'react';
import Task from './components/Task';
import InputTask from './components/InputTask';
import './App.css';

type PropsPerson = {
    firstName: string,
    age: number,
    admin: boolean
};

const App:React.FC = () => {

  const [ btn, setBtn ] = useState<string>("");
  const [change, setChange] = useState<string>("");
  const [ valueToRetrieve, setValueToRetrieve ] = useState<PropsPerson>({
    firstName: "Johnnas",
    age: 44,
    admin: true
  });

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    console.log("clicked !", event.currentTarget);
    setBtn(event.currentTarget.name);
  };

  const handleChangeInput = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setChange(event.target.value);
  };

  const handleValidate = () => {
    setValueToRetrieve({ 
      firstName : change, 
      age: valueToRetrieve.age, 
      admin: valueToRetrieve.admin 
    });
  };

  return (
    <>
      <h1>Hello React</h1>
      
      <Task />

      <InputTask 
        valueToRetrieve={valueToRetrieve} 
        change={change} 
        handleChangeInput={handleChangeInput}
        handleValidate={handleValidate} 
      />

      <button data-testid="btntest" type="button" name="button-one" onClick={handleClick}>
        Click
      </button>
      {btn ? (
        <p>Name of button : {btn}</p> 
        ):(
        <p>Press btn</p>
      )}
    </>
  )
}

export default App
