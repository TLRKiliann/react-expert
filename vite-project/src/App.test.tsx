import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import { it, expect } from "vitest";
import App from "./App";
import InputTask from "./App";
import { handleClick } from "./App";
import { handleValidate } from "./App";

test('renders correctly App', () => {
  const component = renderer.create(<App />)
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

/*
test('renders correctly InputTask', () => {
  const component_2 = renderer.create(
    <InputTask />
  );
  let tree_2 = component_2.toJSON();
  expect(tree_2).toMatchSnapshot();
});
*/

test('testing button by role', () => {
  render(<App />);
  const myBtn = screen.getAllByRole('button');
  expect(myBtn).toHaveLength(2);
});

//function handleClick
test("testing if handleClick function is present", () => {
  const handleClick = vi.fn();
  expect(handleClick).toBeDefined()
});

//button + function
test('handleClick return boolean', () => {const beverage = "yeah";
  const handleClick = vi.fn(beverage => beverage)
  handleClick(beverage)
  expect(handleClick).toHaveReturnedWith("yeah")
  expect(handleClick).toHaveBeenCalledTimes(1)
});

//button + function
test('handleValidate return boolean', () => {const beverage = "yeah";
  const handleValidate = vi.fn(beverage => beverage)
  handleValidate(beverage)
  expect(handleValidate).toHaveReturnedWith("yeah")
  expect(handleValidate).toHaveBeenCalledTimes(1)
});

//tester la valeur de retour
test("should update state with clik", () => {
  const valueTodo = 'ok';
  const handleClick = vi.fn(valueTodo => valueTodo);
  const wrapper = render(
    <button onClick={handleClick} data-testid="btntest">
      Click
    </button>
  );
  const rep = vi.spyOn(React, "useState");
  rep.mockImplementation(size => [size, valueTodo]);
  fireEvent.click(screen.getByTestId('btntest'));
  expect(handleClick).toHaveBeenCalledTimes(1);
  //expect(valueTodo).toHaveReturnedWith('ok')
});

test('handleClick return yeah', () => {
  const beverage = "yeah";
  const handleClick = vi.fn(beverage => beverage)
  handleClick(beverage)
  const { getByTestId } = render(
    <App />
  );
  fireEvent.click(screen.getByTestId('btntest'))
  expect(handleClick).toHaveReturnedWith("yeah")
  expect(handleClick).toHaveBeenCalledTimes(1)
});

test('handleValidate return Cool', () => {
  const beverage = "Cool";
  const handleValidate = vi.fn(beverage => beverage)
  handleValidate(beverage)
  const { getByTestId } = render(
    <InputTask />
  );
  fireEvent.click(screen.getByTestId('btntestvalidate'))
  expect(handleValidate).toHaveReturnedWith("Cool")
  expect(handleValidate).toHaveBeenCalledTimes(1)
});