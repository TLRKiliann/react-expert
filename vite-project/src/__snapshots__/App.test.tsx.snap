// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly App 1`] = `
[
  <h1>
    Hello React
  </h1>,
  <h1>
    Hello Task
  </h1>,
  <h1>
    Hello InputTask
  </h1>,
  <h2
    style={
      {
        "color": "deepskyblue",
      }
    }
  >
    Johnnas
  </h2>,
  <h2
    style={
      {
        "color": "deepskyblue",
      }
    }
  >
    44
  </h2>,
  <h2
    style={
      {
        "color": "deepskyblue",
      }
    }
  >
    Yes admin
  </h2>,
  <label>
    Change firstname :
  </label>,
  <input
    onChange={[Function]}
    type="text"
    value=""
  />,
  <p>
    Not changed
  </p>,
  <button
    data-testid="btntestvalidate"
    onClick={[Function]}
    type="button"
  >
    Validate
  </button>,
  <button
    data-testid="btntest"
    name="button-one"
    onClick={[Function]}
    type="button"
  >
    Click
  </button>,
  <p>
    Press btn
  </p>,
]
`;
